qubes-gui-gntalloc/src/lib.rs

```diff
@@ -0,0 +1,272 @@
+        unsafe {
+            std::slice::from_raw_parts(self.inner.as_ptr() as *const u8, cast_u32(total_length))
+        }
+    }
+}
+
+impl Drop for Buffer {
+    fn drop(&mut self) {
+        let p = ioctl_gntalloc_dealloc_gref {
+            index: self.offset,
+            count: self.dimensions.grefs(),
+        };
+        assert!(self.ptr as usize % 4096 == 0, "Unaligned pointer???");
+        if unsafe { libc::munmap(self.ptr, self.dimensions.buffer_size()) } != 0 {
+            panic!(
+                "the inputs are correct, and this is not punching a hole in an \
+                 existing mapping, so munmap() cannot fail; qed; error {}",
+                io::Error::last_os_error()
+            )
+        }
+        if let Some(alloc) = self.alloc.upgrade() {
+            unsafe {
+                assert_eq!(
+                    libc::ioctl(alloc.as_raw_fd(), IOCTL_GNTALLOC_DEALLOC_GREF, &p),
+                    0,
+                    "Releasing a grant reference never fails; qed",
+                );
+            }
+        } // otherwise, the kernel has done the cleanup when the FD was closed
+    }
+}
+
+#[repr(C)]
+#[allow(nonstandard_style)]
+struct ioctl_gntalloc_alloc_gref {
+    domid: u16,
+    flags: u16,
+    count: u32,
+    index: u64,
+    gref_ids: [u32; 1],
+}
+
+#[repr(C)]
+#[allow(nonstandard_style)]
+struct ioctl_gntalloc_dealloc_gref {
+    index: u64,
+    count: u32,
+}
+
+impl Agent {
+    /// Allocate a buffer to share with the GUI daemon.
+    pub fn alloc_buffer(&mut self, width: u32, height: u32) -> io::Result<Buffer> {
+        let dimensions = dimensions::WindowDimensions::new(width, height)?;
+        assert_eq!(qubes_gui::XC_PAGE_SIZE % 4, 0);
+        let grefs = dimensions.grefs();
+        assert!(
+            grefs <= qubes_gui::MAX_GRANT_REFS_COUNT,
+            "excessive width or height detected above"
+        );
+        let mut channels: Vec<u64> = Vec::with_capacity((cast_u32(grefs) + 5) / 2);
+        unsafe {
+            let ptr = channels.as_mut_ptr() as *mut u8;
+            std::ptr::write(ptr as *mut u16, self.peer);
+            std::ptr::write((ptr as *mut u16).add(1), 1);
+            std::ptr::write((ptr as *mut u32).add(1), grefs);
+            if (grefs & 1) != 0 {
+                assert_eq!(channels.capacity() * 2, cast_u32(grefs) + 5);
+                std::ptr::write((ptr as *mut u32).add(cast_u32(grefs) + 4), 0)
+            } else {
+                assert_eq!(channels.capacity() * 2, cast_u32(grefs) + 4);
+            }
+            let res = libc::ioctl(
+                self.alloc.as_raw_fd(),
+                IOCTL_GNTALLOC_ALLOC_GREF,
+                ptr as *mut ioctl_gntalloc_alloc_gref,
+            );
+            if res != 0 {
+                assert_eq!(res, -1, "invalid return value from ioctl()");
+                return Err(io::Error::last_os_error());
+            }
```

Comment:
Currently you're keeping all your data in the "unused capacity" part of the vector's buffer. This is a supremely weird choice. I'd suggest calling `set_len` here, once the memory is initialized, so that this oddity doesn't become a global concern and is kept to these few lines of the function.

Reply:
Fixed